### PR TITLE
feat: revalidate tag for s3

### PR DESCRIPTION
**Pull request by trsergey25**

### CODE REVIEW

File: packages/cache-core/src/strategies/fileSystem.ts [hunk 1]
```diff
@@ -60,7 +60,7 @@ export class FileSystemCache implements CacheStrategy {
   async deleteAllByKeyMatch(pageKey: string, ctx: CacheContext) {
     if (!existsSync(ctx.serverCacheDirPath)) return
 
-    const pathToCacheFolder = path.join(ctx.serverCacheDirPath, pageKey)
+    const pathToCacheFolder = path.join(ctx.serverCacheDirPath, ...pageKey.split('/'))
```
Comment (RomanBobrovskiy) [inline packages/cache-core/src/strategies/fileSystem.ts:63]: what is the reason for that?

File: packages/s3-cache/src/index.ts [hunk 2]
```diff
@@ -19,6 +20,11 @@ export class S3Cache implements CacheStrategy {
     this.bucketName = bucketName
   }
 
+  buildTagKeys(tags?: string | string[]) {
+    if (!tags || (Array.isArray(tags) && !tags.length)) return ''
```
Comment (RomanBobrovskiy) [inline packages/s3-cache/src/index.ts:24]:
```suggestion
    if (!tags?.length)) return ''
```
it should be enough for both cases with array and string

File: packages/s3-cache/src/index.ts [hunk 3]
```diff
@@ -38,21 +44,58 @@ export class S3Cache implements CacheStrategy {
...
+      for (const { Key: key } of contents) {
+        if (!key) continue
+
+        const args = { Bucket: this.bucketName, Key: key }
+        const { TagSet = [] } = await this.client.getObjectTagging(args)
+
+        const tags = TagSet.filter(({ Key: key }) => key?.startsWith(TAG_PREFIX)).map(({ Value: tags }) => tags || '')
+
+        if (tags.includes(tag)) {
+          await this.client.deleteObject(args)
```
Comment (RomanBobrovskiy) [inline packages/s3-cache/src/index.ts:95]: instead if single deleteObject, let's use `deleteObjects`, so we can delete multiple files via single request, we can send up 1k keys to delete, so we can minimize number of requests what we do to s3

File: packages/cache-core/src/strategies/fileSystem.ts (hunk 1 above)
Comment (trsergey25) [reply packages/cache-core/src/strategies/fileSystem.ts:63]: in the case when we have /page/page/ as a pageKey, and we have to prepare the path for this folder using path.join
for example for Windows OS delimiter is \ and path.join handles it

File: packages/s3-cache/src/index.ts (hunk 2 above)
Comment (trsergey25) [reply packages/s3-cache/src/index.ts]: yes, thank you 
I have fixed it

File: packages/s3-cache/src/index.ts (hunk 3 above)
Comment (trsergey25) [reply packages/s3-cache/src/index.ts]: 😓 Thanks, I reworked it based on `deleteObjects`

File: packages/s3-cache/src/index.ts [hunk 4]
```diff
@@ -79,12 +127,15 @@ export class S3Cache implements CacheStrategy {
         })
       nextContinuationToken = token
 
-      for (const { Key: key } of contents) {
-        if (!key) continue
-        if (key.endsWith('.json') || key.endsWith('.html')) {
-          await this.client.deleteObject({ Bucket: this.bucketName, Key: key })
-        }
-      }
+      const keyToDelete = contents.reduce<string[]>(
```
Comment (RomanBobrovskiy) [inline packages/s3-cache/src/index.ts:130]: I believe we still can optimize it, listObjects also returns up to 1k keys, but not all those keys are going to match to what we need to delete, we should separate that:

1. go through all the objects and collect all keys what we need to delete
2. send deleteObjects request to each 1k of keys

Comment (trsergey25) [reply packages/s3-cache/src/index.ts]: done

**Review by RomanBobrovskiy — APPROVED** [review]

**Pull request merged**